File: cpp/101.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool issym(TreeNode *left, TreeNode *right){
        if(left == nullptr && right == nullptr)
            return true;
        if(left == nullptr || right == nullptr)
            return false;
        return left->val == right->val && issym(left->left, right->right) && issym(left->right, right->left);
    }
    bool isSymmetric(TreeNode *root) {
        if(root == nullptr) return true;
        return issym(root->left, root->right);
    }
};

int main(){
  auto sol = Solution();
}

File: python/101.py
class Solution(object):
    def isSymmetric(self, a, b=None):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if b is None:
            b = a
        if not a and not b:
            return True
        if not a or not b:
            return False
        return a.val == b.val and self.isSymmetric(a.left, b.right) and self.isSymmetric(a.right, b.left)
